clang/labs/lab1/kashirin_alexander/Rename_Plugin.cpp

```diff
@@ -0,0 +1,196 @@
+enum class IdType { Var, Func, Class };
+
+class RenameVisitor : public clang::RecursiveASTVisitor<RenameVisitor> {
+public:
+  explicit RenameVisitor(clang::Rewriter rewriter, IdType type,
+                         clang::StringRef cur_name, clang::StringRef new_name)
+      : rewriter(rewriter), type(type), cur_name(cur_name), new_name(new_name) {
+  }
+
+  bool VisitFunctionDecl(clang::FunctionDecl *func) {
+    if (type == IdType::Func && func->getName() == cur_name) {
+      rewriter.ReplaceText(func->getNameInfo().getSourceRange(), new_name);
+    }
+    return true;
+  }
+
+  bool VisitCallExpr(clang::CallExpr *call) {
+    if (type == IdType::Func) {
+      clang::FunctionDecl *callee = call->getDirectCallee();
+      if (callee && callee->getName() == cur_name) {
+        rewriter.ReplaceText(call->getCallee()->getSourceRange(), new_name);
+      }
+    }
+    return true;
+  }
+
+  bool VisitVarDecl(clang::VarDecl *var) {
+    if (type == IdType::Var && var->getName() == cur_name) {
+      rewriter.ReplaceText(var->getLocation(), cur_name.size(), new_name);
+    }
+    if (type == IdType::Class &&
+        var->getType().getAsString() == cur_name + " *") {
+      rewriter.ReplaceText(var->getTypeSourceInfo()->getTypeLoc().getBeginLoc(),
+                           cur_name.size(), new_name);
+    }
+    if (type == IdType::Class && var->getType().getAsString() == cur_name) {
+      rewriter.ReplaceText(
+          var->getTypeSourceInfo()->getTypeLoc().getSourceRange(), new_name);
+    }
+    return true;
+  }
+
+  bool VisitDeclRefExpr(clang::DeclRefExpr *expr) {
+    clang::VarDecl *var = clang::dyn_cast<clang::VarDecl>(expr->getDecl());
+    if (type == IdType::Var && var && var->getName() == cur_name) {
+      rewriter.ReplaceText(expr->getSourceRange(), new_name);
+    }
+    return true;
+  }
+
+  bool VisitCXXRecordDecl(clang::CXXRecordDecl *record) {
+    if (type == IdType::Class && record->getName() == cur_name) {
+      rewriter.ReplaceText(record->getLocation(), new_name);
+      const auto *destructor = record->getDestructor();
+      if (destructor) {
+        rewriter.ReplaceText(destructor->getLocation(), cur_name.size() + 1,
+                             '~' + new_name);
+      }
+    }
+    return true;
+  }
+
+  bool VisitCXXConstructorDecl(clang::CXXConstructorDecl *constructor) {
+    if (type == IdType::Class) {
+      if (constructor->getNameAsString() == cur_name) {
+        rewriter.ReplaceText(constructor->getLocation(), cur_name.size(),
+                             new_name);
+      }
+    }
+    return true;
+  }
+
+  bool VisitCXXNewExpr(clang::CXXNewExpr *newExpr) {
+    if (type == IdType::Class) {
+      if (newExpr->getConstructExpr()->getType().getAsString() == cur_name) {
+        rewriter.ReplaceText(newExpr->getExprLoc(), cur_name.size() + 4,
+                             "new " + new_name);
+      }
+    }
+    return true;
+  }
+
+  bool save_changes() { return rewriter.overwriteChangedFiles(); }
```

Comment:
Do you need this wrapper function?

Reply:
this function is optional because rewriter.overwriteChangedFiles() can be called directly but save_changes() is shorter